piptools/repositories/pypi.py

```diff
@@ -434,22 +424,23 @@ def _wheel_support_index_min(self: Wheel, tags: List[Tag]) -> int:
 
         original_wheel_supported = Wheel.supported
         original_support_index_min = Wheel.support_index_min
-        original_cache = self._available_candidates_cache
+        original_candidates_cache = self.finder.find_all_candidates
 
         Wheel.supported = _wheel_supported
         Wheel.support_index_min = _wheel_support_index_min
-        self._available_candidates_cache = {}
-
         # If we don't clear this cache then it can contain results from an
         # earlier call when allow_all_wheels wasn't active. See GH-1532
-        self.finder.find_all_candidates.cache_clear()
+        # TODO: Python >= 3.9: Use `lru_cache().cache_parameters()`.
+        self.finder.find_all_candidates = lru_cache(
+            maxsize=self.finder.find_all_candidates.cache_info().maxsize
+        )(self.finder.find_all_candidates.__wrapped__).__get__(self.finder)
 
         try:
             yield
         finally:
             Wheel.supported = original_wheel_supported
             Wheel.support_index_min = original_support_index_min
-            self._available_candidates_cache = original_cache
+            self.finder.find_all_candidates = original_candidates_cache
```

Comment:
A minor note:

`self.finder.find_all_candidates = original_candidates_cache` is not really excellent code IMO, as it doesn't always leave `self.finder` in the same state that `self.finder` was in before `allow_all_wheels(...)` was called. `self.finder.find_all_candidates = original_candidates_cache` adds `{"find_all_candidates": original_candidates_cache}` to `self.finder.__dict__`. For the original `self.finder`, though, `self.finder.__dict__` was empty, and `self.finder.find_all_candidates` would have been resolved from the MRO instead.

`self.finder.find_all_candidates = original_candidates_cache` could be replaced with `del self.finder.find_all_candidates` to avoid polluting `self.finder.__dict__`. Functionally these are the same, but I do think `del` is a whole lot less readable.